README.MD
# FilenameToFiledate

파일의 이름을 파일의 `날짜`로 변경해주는 프로그램

---

## Feature

- 파일의 이름을 파일의 날짜로 변경함
    - 2020년 1월 1일 오후 8시 22분 33초 => `2020_01_01-20_22_33.EXT`
- 프로그램을 작동시킨 디렉토리에 있는 파일들만 이름이 변경됨
- `원본 이름` -> `해시값` -> `수정된 이름` 순으로 변경됨
- 미리 지정한 확장자만 파일 이름이 변경됨
    - 현재 변경되는 확장자 : `.JPG`, `.JPEG`, `.PNG`, `.GIF`, `.HEIC`, `.HEIF`, `.TIF`, `.TIFF`, `.MP4`, `.MOV`, `.AVI`, `.WMV`, `.MPG`, `.MPEG`
- 아래 목록중 가장 오래된 날짜를 기준으로함
    - 사진 : `DateTime`, `DateTimeOriginal`, `DateTimeDigitized`, `Date created`, `Date modified`
    - 비디오 : `Date created`, `Date modified`, `encoded_date`, `tagged_date`
- 중복 네이밍이 발생할 경우, 초단위를 1초씩 올려서 중복 해결
    - `2020_02_02-22_22_01.EXT` => `2020_02_02-22_22_01.EXT`
    - `2020_02_02-22_22_01.EXT` => `2020_02_02-22_22_02.EXT`
    - `2020_02_02-22_22_02.EXT` => `2020_02_02-22_22_03.EXT`
    - `2020_02_02-22_22_02.EXT` => `2020_02_02-22_22_04.EXT`
    - `2020_02_02-22_22_04.EXT` => `2020_02_02-22_22_05.EXT`

---

## Caution

- 프로그램 실행 시 오류가 뜨게 된다면, `바이러스 및 위협 방지`에 들어가서 이 프로그램을 `디바이스에서 허용` 해주어야함
- 프로그램 실행 후 이미 `바뀐 파일 이름은 되돌릴 수 없음`
- 확장자가 소문자일 경우 `대문자`로 변경됨

---

## Extra Tools

- [hour_converter.py](https://github.com/syki66/filename_to_filedate/blob/master/extra_tools/hour_converter.py) : 시간대를 변경해주는 툴
- [second_converter.py](https://github.com/syki66/filename_to_filedate/blob/master/extra_tools/second_converter.py) : 초단위를 변경해주는 툴
- [line_up_seconds.py](https://github.com/syki66/filename_to_filedate/blob/master/extra_tools/line_up_seconds.py) : 기준값 날짜로 줄세워주는 툴
- [timestamp_to_date_converter.py](https://github.com/syki66/filename_to_filedate/blob/master/extra_tools/timestamp_to_date_converter.py) : 타임스탬프를 날짜로 변환해주는 툴

---

## Change Log

- [CHANGELOG.MD](https://github.com/syki66/filename_to_timestamp/blob/master/CHANGELOG.MD)

---

## Todo List

- `heic`, `heif`의 exif를 자체적으로 해결하기
- gui 환경 구성하면서 보조툴 기능 병합
- 코드 분리하기
- 원본 파일이름 남길지 선택 옵션
- YYYY MM DD 이런것들 사이에 넣을수 있는 문자를 커스터마이징 할수 있게하기

---

## Build

> pyinstaller 사용시, 같은 디렉토리에 `filename_to_datetime.py`, `exiftool.exe`, `pen.ico` 위치시키고, 및 `pymediainfo`에 있는 dll 파일 위치 넣어주기

```
pyinstaller --onefile --icon pen.ico --add-data "exiftool.exe;." --add-binary "C:\Users\syki66\AppData\Local\Programs\Python\Python39\Lib\site-packages\pymediainfo\MediaInfo.dll;." filename_to_datetime.py
```

CHANGELOG.MD
Version 0.2.15 (build 2021-01-30)

- 일부 확장자가 리네이밍 되지 않는 버그 수정

---

Version 0.2.14 (build 2021-01-26)

- 보조툴 추가 *(`second_converter.py`)*
    - `second_converter.py` : 초를 입력하게 되면, 파일 이름에서 입력 초만큼 날짜가 계산된 후, 리네이밍

---

Version 0.2.13 (build 2021-01-15)

- 확장자 리스트 수정
    - 추가 : `.WMV`, `.MPG`, `.MPEG`
    - 제거 : `.JPS`, `.K3G`

---

Version 0.2.12 (build 2021-01-03)

- 보조툴 추가 *(`timestamp_to_date_converter.py`)*
    - `timestamp_to_date_converter.py` : 타임스탬프를 가진 파일이름을 날짜값으로 변경해줌
- 보조툴 수정 *(`hour_converter.py`, `line_up_seconds.py`)*
    - 변환시 확장자를 대문자로 바꿈

---

Version 0.2.11 (build 2021-01-01)

- 보조툴 수정 *(`line_up_seconds.py`)*
    - `line_up_seconds.py` : second 변수의 초기값도 사용자가 직접 입력할 수 있도록 변경

---

Version 0.2.10 (build 2020-12-25)

- 보조툴 수정 *(`line_up_seconds.py`)*
    - `line_up_seconds.py` : 초단위가 99를 초과하게되면 분단위를 1 증가시키도록 변경

---

Version 0.2.9 (build 2020-12-05)

- 보조툴 2가지 추가 *(`hour_converter.py`, `line_up_seconds.py`)*
    - `hour_converter.py` : 시간을 입력하게 되면, 파일 이름에서 입력 시간만큼 날짜가 계산된 후, 리네이밍
    - `line_up_seconds.py` : 수동으로 `YYYY-MM-DD_hh-mm` 형식을 입력하면, 초단위를 00부터 1씩 올리며 리네이밍

---

Version 0.2.8 (build 2020-12-04)

- HEIF(.heic) 파일의 exif 추출 문제 때문에 [exiftool.exe](https://exiftool.org/)을 `FilenameToFiledate.exe` 내부에 집어 넣음
- "준비중" 문구 추가 및 콘솔에 뜨는 문구 일부 수정

---

Version 0.2.7 (build 2020-12-03)

- `exiftool.exe` 파일이 동일 디렉토리에 존재한다면 heic, heif 파일의 exif 정보를 불러올 수 있음
- 변환될 파일 개수를 미리 보여줌

---

Version 0.2.6 (build 2020-12-02)

- 큰 이미지 파일 불러올때 뜨는 `DecompressionBombWarning` 경고 제거
- 비디오 메타데이터 날짜 2종류 추가 (`encoded_date`, `tagged_date`)

---

Version 0.2.5 (build 2020-12-01)

- 미디어 파일 확장자 추가 *(`.TIF`, `.AVI`, `.MP4`, `.MOV`, `.K3G`, `.JPS`)*
- exif 데이터에 들어있는 날짜정보 3개와 파일날짜정보 2개 중에 가장 오래된 날짜를 기준으로 이름변경하도록 함
- 프로그램 시작시 경고 메세지 및 실행되는 디렉토리 경로 표시

---

Version 0.2.3 (build 2020-10-28)

- 파일이름이 이미 존재하더라도 에러가 안나도록 변경 *(해시값 사용)*

---

Version 0.2.2 (build 2020-10-27)

- 사용자를 위한 시각적 로그 추가 (변환된 파일 개수 출력, 중복된 파일이름 처리 현황 보여주기)

---

Version 0.2.1 (build 2020-10-27)

- 중복 네이밍이 발생할 경우, 초단위를 1초씩 올려서 중복 해결 (뒤에 파일 날짜를 침범할 경우 뒤에 있는 기존 파일의 초단위도 같이 밀어버림)

---

Version 0.2.0 (build 2020-10-26)

- 기존 방식에서, 파일의 중복된 날짜 처리 이슈 때문에, 같은 디렉토리의 파일들을 모두 불러와서 한번에 처리하는 방식으로 변경
- 중첩 사전형을 이용하여 확장자 별로 먼저 분리하고, 각 확장자별로 날짜를 기준으로 오름차순으로 정렬된 사전형을 사용함

---

Version 0.1.1 (build 2020-10-22)

- 변환된 개수 출력해주기
- 확장자가 소문자일 경우 대문자로 변경됨
- `찍힌 날짜`가 존재하지 않을경우 `수정한 날짜` 값으로 이름 변경

---

Version 0.1.0 (build 2020-10-21)

- 파일을 하나씩 불러와서 이름 바꾸는 방식
- 파일의 이름을 파일의 수정한 날짜로 변경함
- 이 파일을 실행하면 같은 디렉토리에 있는 파일들만 이름이 변경됨
- 미리 지정한 확정자만 이름이 변경되도록 함 *(현재는 이미지 파일들만 변경되게함)*
- 확장자가 대문자일 경우 소문자로 변경됨
- 사용자가 변경된 파일을 알 수 있도록 로그를 출력
- 프로그램 실행시 파일명을 변경할 건지 다시 한번 확인

exif_test.py
from PIL import Image
import os
from datetime import datetime

#exif 변환해야됨

# get "Date created", "Date modified"
def getFileDate(filename):
    dateArray = []
    if os.path.getctime(filename) is not None:
        timestamp = os.path.getctime(filename)
        date = datetime.fromtimestamp(timestamp)
        dateCreated = f'{date.year:04}-{date.month:02}-{date.day:02}_{date.hour:02}-{date.minute:02}-{date.second:02}'
        dateArray.append(dateCreated)
    if os.path.getmtime(filename) is not None:
        timestamp = os.path.getmtime(filename)
        date = datetime.fromtimestamp(timestamp)
        dateModified = f'{date.year:04}-{date.month:02}-{date.day:02}_{date.hour:02}-{date.minute:02}-{date.second:02}'
        dateArray.append(dateModified)
    return dateArray

# get Dates of exif data ("DateTimeOriginal", "DateTimeDigitized", "DateTime")
def getExifDate(filename):
    image = Image.open(filename)
    exif = image.getexif()

    dateArray = []
    if 36867 in exif:
        dateTimeOriginal = f'{int(exif[36867][0:4]):04}-{int(exif[36867][5:7]):02}-{int(exif[36867][8:10]):02}_{int(exif[36867][11:13]):02}-{int(exif[36867][14:16]):02}-{int(exif[36867][17:19]):02}'
        dateArray.append(dateTimeOriginal)
    if 36868 in exif:
        dateTimeDigitized = f'{int(exif[36868][0:4]):04}-{int(exif[36868][5:7]):02}-{int(exif[36868][8:10]):02}_{int(exif[36868][11:13]):02}-{int(exif[36868][14:16]):02}-{int(exif[36868][17:19]):02}'
        dateArray.append(dateTimeDigitized)
    if 306 in exif:
        dateTime = f'{int(exif[306][0:4]):04}-{int(exif[306][5:7]):02}-{int(exif[306][8:10]):02}_{int(exif[306][11:13]):02}-{int(exif[306][14:16]):02}-{int(exif[306][17:19]):02}'
        dateArray.append(dateTime)
    return dateArray

def pickOldestDate(filename):
    return getFileDate(filename) + getExifDate(filename)



for file in os.listdir():


    try:
        print(file)
        print(pickOldestDate(file))
        print("")

    except Exception as e:

        print(f'errorName:{e}')



# imagename = "IMG_2620.JPG"
# print(pickOldestDate(imagename))






'''
DateTime
DateTimeOriginal
DateTimeDigitized
Date created
Date modified
'''


metadata_test.py
from pymediainfo import MediaInfo
from datetime import datetime
from dateutil import tz

media_info = MediaInfo.parse("test.mp4")

timestring = media_info.tracks[2].to_data()['encoded_date'].split('UTC ')[1]
print(media_info.tracks[2].to_data()['encoded_date'].split('UTC ')[1])


from_zone = tz.tzutc()
to_zone = tz.tzlocal()

utc = datetime.strptime(timestring, '%Y-%m-%d %H:%M:%S')
utc = utc.replace(tzinfo=from_zone)

central = utc.astimezone(to_zone)

print(central)

HEIC_exif_test.py
import subprocess

input_file = "test.HEIC"
exe = "exiftool"
process = subprocess.Popen([exe, input_file], stdout=subprocess.PIPE, stderr=subprocess.STDOUT, universal_newlines=True)

info = {}
for output in process.stdout:
    
    line = output.strip().split(":",1)
    info[line[0].strip()] = line[1].strip()
    print(line)

print(info['Date/Time Original'])
print(info['Create Date'])
print(info['Modify Date'])

filename_to_timestamp.py
from datetime import datetime
import os
from collections import OrderedDict
import hashlib

extensions = [".JPG", ".JPEG", ".PNG", ".GIF", ".HEIC", "HEIF"]

# 파일의 날짜 얻기, (찍은 날짜 존재하면 찍은 날짜로 대입하는 것으로 변경해야됨)
def getFileDate(filename):
    timestamp = os.path.getmtime(filename)
    date = datetime.fromtimestamp(timestamp)
    
    new_name = f'{date.year:04}-{date.month:02}-{date.day:02}_{date.hour:02}-{date.minute:02}-{date.second:02}'
    return new_name

# 찍은 날짜가 존재하는지 함수 추가하기
# heic 파일은 jpg 변환후 찍은날짜 가져와야됨

# 각 확장자마다, 날짜기준 오름차순으로 만든 OrderedDict형을 반환함
def sortFiles():
    files = {}
    for ext in extensions:
        temp = {}
        for filename in os.listdir():
            if (filename[filename.rfind('.'):].upper() == ext):
                temp[filename] = f'{getFileDate(filename)}{ext}'
        files[ext] = OrderedDict(sorted(temp.items(),key=lambda x: x[1]))
    return files

# 파일명 재조립하기
def reassemble(value, number):
    date = value[ :value.rfind('-')]
    second = int(value[value.rfind('-')+1 : value.rfind('.')])
    extension = value[value.rfind('.') : ]
    return f'{date}-{second+number:02}{extension}'

# 날짜가 중복된 파일에만 중복이 아닐때까지 초단위 시간 올리기
def handleDuplicates(dict):
    for ext in extensions:
        prev_values = []
        for key, value in dict[ext].items():
            dict_without_me = { k:v for k, v in dict[ext].items() if ( k != key) }
            num = 0
            # 중복파일이 존재하고 이전에 존재했던 값이 아닐때까지 1씩 계속 증가시키기
            while (reassemble(value, num) in dict_without_me.values() and reassemble(value, num) in prev_values ):
                num += 1
            dict[ext][key] = reassemble(value, num)
            prev_values.append(reassemble(value, num))
    return dict

# 파일이름 변경 추적 현황을 콘솔에 출력해주기
original_dict = sortFiles()
def printFilenameStatus(extension, key, value):
    if (original_dict[extension][key] != value):
        print(f'CHANGED: "{key}" -> "{original_dict[extension][key]}" -> "{value}"')
    else:
        print(f'CHANGED: "{key}" -> "{value}"')

# 파일 이름을 해시값으로 변환
def convertToHash(filename):
    encoded_name = filename.encode('utf-8')
    md5 = hashlib.new('md5')
    md5.update(encoded_name)
    hash_value = md5.hexdigest()
    return hash_value

# 파일 이름을 전부 해시값으로 변환 (중복 이름 에러 방지)
def wipeUpFileName(dict):
    for ext in extensions:
        for key, value in dict[ext].items():
            hashed_key = convertToHash(key)
            os.rename(key, hashed_key)

# 실제 파일이름 변경
converted_file_count = 0
def renameFiles(dict):
    global converted_file_count
    for ext in extensions:
        for key, value in dict[ext].items():
            hashed_key = convertToHash(key)
            os.rename(hashed_key, value)
            printFilenameStatus(ext, key, value)
            converted_file_count += 1

# 실행부
user_input = input("IF YOU TYPE \"YES\" OR \"Y\", THE IMAGE FILES ARE RENAMED TO TIMESTAMP\n")
yes = True if (user_input.lower() == 'y' or user_input.lower() == 'yes') else False

try:
    if (yes):
        print("\nCHANGING IMAGE FILES NAME...\n")

        refined_dict = handleDuplicates(sortFiles())
        wipeUpFileName(refined_dict)
        renameFiles(refined_dict)

        if converted_file_count == 0:
            print("NO FILES ARE CONVERTED")
        else:
            print(f'\nFINISHED. CONVERTED {converted_file_count} FILES')
    else:
        print("GOOD BYE")
except:
    print("ERROR")

input("\nPRESS ENTER TO EXIT.")

extra_tools/hour_converter.py
from datetime import datetime, timedelta
import os

print('주의 : 같은 디렉토리에 있는 모든 파일이 변환 과정을 거침\n')
print('확장자는 대문자로 변경됨')
print('\"+\"숫자 또는 \"-숫자\"를 이용해서 시간값만 변경가능')
print('9시간 이전으로 이름변경을 한다면 \"-9\"를 입력')
print('\"2020-01-01_11-11-23.ext\" 이 형식일 경우만 변환됨\n')

def changeHour(hour):
    for old_filename in os.listdir():
        try:
            file_date = old_filename.split(".")[0]
            ext = old_filename.split(".")[1].upper()

            old_date = datetime.strptime(file_date, "%Y-%m-%d_%H-%M-%S")
            new_date= old_date + timedelta(hours=hour)
            new_filename = f'{new_date.year:04}-{new_date.month:02}-{new_date.day:02}_{new_date.hour:02}-{new_date.minute:02}-{new_date.second:02}.{ext}'

            os.rename(old_filename, new_filename)
            print(f'변경됨 : {old_filename} -> {new_filename}')
        except Exception as e:
            print(f'에러발생 : \"{old_filename}\" {e}')

try:
    hour = int(input("시간 값을 입력해주세요 : "))
    changeHour(hour)
except Exception as e:
    print(f'에러발생 : {e}')

input("\n엔터를 누르면 종료됨")

extra_tools/line_up_seconds.py
import os
import re

print('주의 : 같은 디렉토리에 있는 모든 파일이 변환 과정을 거침\n')
print('확장자는 대문자로 변경됨')
print('\"YYYY-MM-DD_hh-mm-ss.ext\" 이 형식의 이름을 가진 파일만 변환됨')
print('입력은 \"YYYY-MM-DD_hh-mm-ss\" 형식으로 입력해야됨')
print('입력 날짜 기준으로 초단위 오름차순으로 파일의 이름을 변경함')
print('초단위가 99를 초과할때 분단위가 1 올라감\n')

def isValidValue(value, regex, maxLength):
    if len(value) > maxLength:
        raise Exception('형식이 맞지 않습니다.')
    else:
        p = re.compile(regex)
        m = p.match(value)
        if m is not None:
            return True
        else:
            raise Exception('형식이 맞지 않습니다.')

def changeName(date):
    new_date = date.rsplit('-', 2)[0]
    minute = int(date.rsplit('-', 2)[1])
    second = int(date.rsplit('-', 2)[2])

    for file in os.listdir():
        try:
            old_filename = file
            old_name = old_filename.split(".")[0]
            ext = old_filename.split(".")[1].upper()
            if isValidValue(old_name, '\d\d\d\d-\d\d-\d\d_\d\d-\d\d-\d\d', 19):
                new_filename = f'{new_date}-{minute:02}-{second:02}.{ext}'
                os.rename(old_filename, new_filename)
                print(f'변경됨 : {old_filename} -> {new_filename}')
                if (second >= 99):
                    minute += 1
                    second = 0
                else:
                    second += 1
        except Exception as e:
            print(f'에러발생 : \"{file}\" 파일의 {e} \"YYYY-MM-DD_hh-mm-ss.ext\"')

try:
    date = input("시간 값을 입력해주세요 : ")
    if (isValidValue(date, '\d\d\d\d-\d\d-\d\d_\d\d-\d\d-\d\d', 19)):
        changeName(date)
except Exception as e:
    print(f'\n에러발생 : 입력값의 {e} \"YYYY-MM-DD_hh-mm-ss\"')

input("\n엔터를 누르면 종료됨")


extra_tools/second_converter.py
from datetime import datetime, timedelta
import os

print('주의 : 같은 디렉토리에 있는 모든 파일이 변환 과정을 거침\n')
print('확장자는 대문자로 변경됨')
print('\"+\"숫자 또는 \"-숫자\"를 이용해서 초 값만 변경가능')
print('3분 이전으로 이름변경을 한다면 \"-180\"를 입력')
print('\"2020-01-01_11-11-23.ext\" 이 형식일 경우만 변환됨\n')

def changeSecond(second):
    for old_filename in os.listdir():
        try:
            file_date = old_filename.split(".")[0]
            ext = old_filename.split(".")[1].upper()

            old_date = datetime.strptime(file_date, "%Y-%m-%d_%H-%M-%S")
            new_date= old_date + timedelta(seconds=second)
            new_filename = f'{new_date.year:04}-{new_date.month:02}-{new_date.day:02}_{new_date.hour:02}-{new_date.minute:02}-{new_date.second:02}.{ext}'

            os.rename(old_filename, new_filename)
            print(f'변경됨 : {old_filename} -> {new_filename}')
        except Exception as e:
            print(f'에러발생 : \"{old_filename}\" {e}')

try:
    second = int(input("초 값을 입력해주세요 : "))
    changeSecond(second)
except Exception as e:
    print(f'에러발생 : {e}')

input("\n엔터를 누르면 종료됨")

filename_to_datetime.py
from datetime import datetime
import os
from collections import OrderedDict
import hashlib
from PIL import Image
from pymediainfo import MediaInfo
from dateutil import tz

extensions = [".JPG", ".JPEG", ".PNG", ".GIF", ".HEIC", ".HEIF", ".TIF", ".TIFF", ".MP4", ".MOV", ".AVI", ".WMV", ".MPG", ".MPEG"]

print("PREPARING FOR RENAMING FILES...\n")

# get exif of HEIF HEIC image (heic 자체 지원 방법 찾으면 나중에 삭제할 예정)
def getHeicExif(filename):
    import subprocess
    dateArray = []
    def getDate(value):
        date = str(datetime.strptime(value, '%Y:%m:%d %H:%M:%S'))
        return f'{int(date[0:4]):04}-{int(date[5:7]):02}-{int(date[8:10]):02}_{int(date[11:13]):02}-{int(date[14:16]):02}-{int(date[17:19]):02}'
    def resource_path(relative_path):
        """ Get absolute path to resource, works for dev and for PyInstaller """
        try:
            # PyInstaller creates a temp folder and stores path in _MEIPASS
            base_path = sys._MEIPASS
        except Exception:
            base_path = os.path.abspath(".")
        return os.path.join(base_path, relative_path)
    try:
        if filename.upper().endswith('.HEIC') or filename.upper().endswith('.HEIF'):
            exe = resource_path("exiftool.exe")
            process = subprocess.Popen([exe, filename], stdout=subprocess.PIPE, stderr=subprocess.STDOUT, universal_newlines=True)
            exif = {}
            for output in process.stdout:
                line = output.strip().split(":",1)
                exif[line[0].strip()] = line[1].strip()
            if 'Date/Time Original' in exif:
                dateArray.append(getDate(exif['Date/Time Original'].split('.')[0]))
            if 'Create Date' in exif:
                dateArray.append(getDate(exif['Create Date'].split('.')[0]))
            if 'Modify Date' in exif:
                dateArray.append(getDate(exif['Modify Date'].split('+')[0]))
    except:
        pass
    return dateArray

# get Dates of video metadata ("encoded_date", "tagged_date")
def getVideoDate(filename):
    dateArray = []
    def getDate(key):
        value = media_info.tracks[2].to_data()[key].split('UTC ')[1]
        utcDate = datetime.strptime(value, '%Y-%m-%d %H:%M:%S')
        utcDate = utcDate.replace(tzinfo=tz.tzutc())
        localDate = str(utcDate.astimezone(tz.tzlocal())).split('+')[0]
        return f'{int(localDate[0:4]):04}-{int(localDate[5:7]):02}-{int(localDate[8:10]):02}_{int(localDate[11:13]):02}-{int(localDate[14:16]):02}-{int(localDate[17:19]):02}'
    try:
        media_info = MediaInfo.parse(filename)
        if 'encoded_date' in media_info.tracks[2].to_data():
            dateArray.append(getDate('encoded_date'))
        if 'tagged_date' in media_info.tracks[2].to_data():
            dateArray.append(getDate('tagged_date'))
    except:
        pass
    return dateArray

# get "Date created", "Date modified"
def getFileDate(filename):
    dateArray = []
    if os.path.getctime(filename) is not None:
        timestamp = os.path.getctime(filename)
        date = datetime.fromtimestamp(timestamp)
        dateCreated = f'{date.year:04}-{date.month:02}-{date.day:02}_{date.hour:02}-{date.minute:02}-{date.second:02}'
        dateArray.append(dateCreated)
    if os.path.getmtime(filename) is not None:
        timestamp = os.path.getmtime(filename)
        date = datetime.fromtimestamp(timestamp)
        dateModified = f'{date.year:04}-{date.month:02}-{date.day:02}_{date.hour:02}-{date.minute:02}-{date.second:02}'
        dateArray.append(dateModified)
    return dateArray

# get Dates of exif data ("DateTimeOriginal", "DateTimeDigitized", "DateTime")
def getExifDate(filename):
    dateArray = []
    Image.MAX_IMAGE_PIXELS = None # DecompressionBombWarning 안뜨게 하기
    try:
        image = Image.open(filename)
        exif = image.getexif()
        if 36867 in exif:
            dateTimeOriginal = f'{int(exif[36867][0:4]):04}-{int(exif[36867][5:7]):02}-{int(exif[36867][8:10]):02}_{int(exif[36867][11:13]):02}-{int(exif[36867][14:16]):02}-{int(exif[36867][17:19]):02}'
            dateArray.append(dateTimeOriginal)
        if 36868 in exif:
            dateTimeDigitized = f'{int(exif[36868][0:4]):04}-{int(exif[36868][5:7]):02}-{int(exif[36868][8:10]):02}_{int(exif[36868][11:13]):02}-{int(exif[36868][14:16]):02}-{int(exif[36868][17:19]):02}'
            dateArray.append(dateTimeDigitized)
        if 306 in exif:
            dateTime = f'{int(exif[306][0:4]):04}-{int(exif[306][5:7]):02}-{int(exif[306][8:10]):02}_{int(exif[306][11:13]):02}-{int(exif[306][14:16]):02}-{int(exif[306][17:19]):02}'
            dateArray.append(dateTime)
    # except Exception as e:
    #     print(e)
    except:
        pass
    return dateArray

# 가장 오래된 날짜만 반환해주는 함수
def pickOldestDate(filename):
    date = getFileDate(filename) + getExifDate(filename) + getVideoDate(filename) + getHeicExif(filename) # (heic, 삭제 예정)
    date.sort()
    return date[0]

# 각 확장자마다, 날짜기준 오름차순으로 만든 OrderedDict형을 반환함
def sortFiles():
    files = {}
    for ext in extensions:
        temp = {}
        for filename in os.listdir():
            if (filename[filename.rfind('.'):].upper() == ext):
                temp[filename] = f'{pickOldestDate(filename)}{ext}'
        files[ext] = OrderedDict(sorted(temp.items(),key=lambda x: x[1]))
    return files

# 파일명 재조립하기
def reassemble(value, number):
    date = value[ :value.rfind('-')]
    second = int(value[value.rfind('-')+1 : value.rfind('.')])
    extension = value[value.rfind('.') : ]
    return f'{date}-{second+number:02}{extension}'

# 날짜가 중복된 파일에만 중복이 아닐때까지 초단위 시간 올리기
def handleDuplicates(dict):
    for ext in extensions:
        prev_values = []
        for key, value in dict[ext].items():
            dict_without_me = { k:v for k, v in dict[ext].items() if ( k != key) }
            num = 0
            # 중복파일이 존재하고 이전에 존재했던 값이 아닐때까지 1씩 계속 증가시키기
            while (reassemble(value, num) in dict_without_me.values() and reassemble(value, num) in prev_values ):
                num += 1
            dict[ext][key] = reassemble(value, num)
            prev_values.append(reassemble(value, num))
    return dict

# 파일이름 변경 추적 현황을 콘솔에 출력해주기
original_dict = sortFiles()
def printFilenameStatus(extension, key, value):
    if (original_dict[extension][key] != value):
        print(f'CHANGED: "{key}" -> "{original_dict[extension][key]}" -> "{value}"')
    else:
        print(f'CHANGED: "{key}" -> "{value}"')

# 파일 이름을 해시값으로 변환
def convertToHash(filename):
    encoded_name = filename.encode('utf-8')
    md5 = hashlib.new('md5')
    md5.update(encoded_name)
    hash_value = md5.hexdigest()
    return hash_value

# 파일 이름을 전부 해시값으로 변환 (중복 이름 에러 방지)
def wipeUpFileName(dict):
    for ext in extensions:
        for key, value in dict[ext].items():
            hashed_key = convertToHash(key)
            os.rename(key, hashed_key)

# 실제 파일이름 변경
converted_file_count = 0
def renameFiles(dict):
    global converted_file_count
    for ext in extensions:
        for key, value in dict[ext].items():
            hashed_key = convertToHash(key)
            os.rename(hashed_key, value)
            printFilenameStatus(ext, key, value)
            converted_file_count += 1

# 변환전 변환될 파일 개수 출력해주기
def ToBeConvertedCount(dict):
    count = 0
    for ext in extensions:
        count += len(dict[ext])
    return count

# 실행부
print('WARNING!! THIS OPERATION IS IRREVRERSIBLE!!\n')
print(f'YOUR CURRENT DIRECTORY IS \"{os.getcwd()}\"\n')
print(f'{ToBeConvertedCount(original_dict)} MEDIA FILES WILL BE RENAMED.\n')

user_input = input("IF YOU TYPE \"YES\" OR \"Y\", THE MEDIA FILES ARE RENAMED TO \"ORIGINAL FILE DATE\"\n>>> ")
yes = True if (user_input.lower() == 'y' or user_input.lower() == 'yes') else False

try:
    if (yes):
        print("\nCHANGING MEDIA FILES NAME...\n")

        refined_dict = handleDuplicates(sortFiles())
        wipeUpFileName(refined_dict)
        renameFiles(refined_dict)

        if converted_file_count == 0:
            print("NO FILES ARE RENAMED")
        else:
            print(f'\nFINISHED. {converted_file_count} FILES RENAMED')
    else:
        print("GOOD BYE")
except Exception as e:
    print("ERROR")
    print(e)

input("\nPRESS ENTER TO EXIT.")
